File: backend/models/CourseModel.py
import sys
from .db import db
import json

from sqlalchemy_utils import JSONType, ChoiceType
from enum import Enum
from models.SpecializationModel import Specialization


class Requirement(db.Model):
    __tablename__ = 'requirement'
    id = db.Column(db.Integer, primary_key=True)
    parent_course_id = db.Column(db.Integer, db.ForeignKey('course.id'), nullable=False)
    child_course_id = db.Column(db.Integer, db.ForeignKey('course.id'), nullable=False)
    req_type = db.Column(db.String(80), nullable=False)
    parent_course = db.relationship('Course', backref='requisite_parent', foreign_keys=[parent_course_id])
    child_course = db.relationship('Course', backref='requisite_child', foreign_keys=[child_course_id])

COURSE_CHOICES = [
    ('1', 'Fall Only'),
    ('2', 'Winter Only'),
    ('3', 'Spring/Summer Only'),
    ('4', 'Fall and Winter'),
    ('5', 'Fall and Spring/Summer'),
    ('6', 'Winter and Spring/Summer'),
    ('7', 'All Semesters')
]

class Course(db.Model):
    __tablename__ = 'course'
    id = db.Column(db.Integer, primary_key=True)
    course_profile = db.Column(db.String(150), nullable=False)
    number_of_credits = db.Column(db.Integer, nullable=False)
    course_description = db.Column(db.Text)
    course_number = db.Column(db.String(4), nullable=False)
    department = db.Column(db.String(3), nullable=False)
    active = db.Column(db.Boolean, default=False)
    ms_code = db.Column(ChoiceType(COURSE_CHOICES), nullable=True)

File: frontend/src/app/shared/models/course.model.ts
import { PowCourseModel } from './pow.model';

export class CourseModel {
  id: number | undefined;
  courseProfile: string | undefined;
  numberOfCredits: number | undefined;
  courseDescription: string | undefined;
  courseNumber: string | undefined;
  department: string | undefined;
  active: boolean | undefined;
  masterScheduleCode: number | undefined;
  requisites: RequisiteModel[] | undefined;

  constructor(courseModel: { [key: string]: any } | undefined) {
    this.id = courseModel?.id;
    this.courseProfile = courseModel?.course_profile;
    this.numberOfCredits = courseModel?.number_of_credits;
    this.courseDescription = courseModel?.course_description;
    this.courseNumber = courseModel?.course_number;
    this.department = courseModel?.department;
    this.active = courseModel?.active;
    this.masterScheduleCode = courseModel?.ms_code;
    this.requisites = courseModel?.requisite_parent?.map(
      (course: { [key: string]: any } | undefined) =>
        new RequisiteModel({ ...course })
    );
  }
}

type ReqType = 'co' | 'pre';

export class RequisiteModel {
  id: number | undefined;
  parentCourse: CourseModel | undefined;
  childCourse: CourseModel | undefined;
  reqType: ReqType | undefined;

  constructor(requisite: { [key: string]: any } | undefined) {
    this.id = requisite?.id;
    this.parentCourse = new CourseModel({ ...requisite?.parent_course });
    this.childCourse = new CourseModel({ ...requisite?.child_course });
    this.reqType = requisite?.req_type;
  }
}
